(GH-1868) Add support for downloading files from targets (#1979)
puppetlabs/bolt: documentation/running_bolt_commands.md

    ```
    bolt file upload my_file.txt /tmp/remote_file.txt --targets web5.mydomain.edu,web6.mydomain.edu
    ```

## Download files or directories from remote targets

Use Bolt to copy files or directories from remote targets to your local system.

To download a file or directory from a remote target, run the
`bolt file download` command. Specify the remote path to the file or
directory, the destination directory on your local system, and the targets.

```shell
$ bolt file download <SOURCE> <DESTINATION> --targets <TARGETS>
```

The `destination` can be either an absolute or relative path to a directory

> on your local system. If a relative path is given, the path is expanded
> relative to the current working directory. If the destination directory does
> not exist, Bolt will automatically create it.

Reviewer: ```suggestion
on your local system. If you use a relative path, Bolt expands the path
relative to the current working directory. If the destination directory does
not exist, Bolt will automatically create it.
```

    ```
    bolt file upload my_file.txt /tmp/remote_file.txt --targets web5.mydomain.edu,web6.mydomain.edu
    ```

## Download files or directories from remote targets

Use Bolt to copy files or directories from remote targets to your local system.

To download a file or directory from a remote target, run the
`bolt file download` command. Specify the remote path to the file or
directory, the destination directory on your local system, and the targets.

```shell
$ bolt file download <SOURCE> <DESTINATION> --targets <TARGETS>
```

The `destination` can be either an absolute or relative path to a directory

> on your local system. If a relative path is given, the path is expanded

Reviewer: Just double-checking something: `downloads` is only used when you use a relative path? So if I gave an absolute path, I could specify any directory name and the result would not be nested under a `downloads` directory?

Author: Correct

Reviewer: Okay cool, I think this bit is good then 👍

relative to the `downloads` directory in the [project 
directory](bolt_project_directories.md), for example
`~/.puppetlabs/bolt/downloads`.

Each file or directory is saved to the destination directory under a
directory with a name matching the URL-encoded name of the target it
was downloaded from. The target directory names are URL-encoded to ensure
that they are valid directory names.

> For example, running the following command will download the SSH daemon
> configuration file from each target:

Reviewer: ```suggestion
For example, the following command downloads the SSH daemon configuration file from two targets, `linux` and `ssh://example.com`:
```

```shell
$ bolt file download /etc/ssh/sshd_config sshd_config --targets linux,ssh://example.com
```

> After running this command, your project directory structure would look

Reviewer: ```suggestion
After running this command from the root of your project directory, your project directory structure would look
```

> like:

Reviewer: ```suggestion
like this:
```

```shell
$ tree
.
├── bolt-project.yaml
├── inventory.yaml
└── downloads/
    └── sshd_config/
        ├── linux/
        │   └── sshd_config
        └── ssh%3A%2F%2Fexample.com/
            └── sshd_config
```

> 🔩 **Tip:** To avoid URL encoding the target's safe name, give the target a

> > simple, human-readable name in your inventory file.

Reviewer: This is a great tip!
